source/common/config/context_provider_impl.h

```diff
@@ -13,10 +14,36 @@ class ContextProviderImpl : public ContextProvider {
                       const Protobuf::RepeatedPtrField<std::string>& node_context_params)
       : node_context_(XdsContextParams::encodeNodeContext(node, node_context_params)) {}
 
+  // Config::ContextProvider
   const xds::core::v3::ContextParams& nodeContext() const override { return node_context_; }
+  const xds::core::v3::ContextParams&
+  dynamicContext(absl::string_view resource_type_url) const override {
+    auto it = dynamic_context_.find(resource_type_url);
+    if (it != dynamic_context_.end()) {
+      return it->second;
+    }
+    return xds::core::v3::ContextParams::default_instance();
+  };
+  void setDynamicContextParam(absl::string_view resource_type_url, absl::string_view key,
+                              absl::string_view value) override {
+    (*dynamic_context_[resource_type_url].mutable_params())[key] = value;
+    update_cb_helper_.runCallbacks(resource_type_url);
+  }
+  void unsetDynamicContextParam(absl::string_view resource_type_url,
+                                absl::string_view key) override {
+    dynamic_context_[resource_type_url].mutable_params()->erase(key);
+    update_cb_helper_.runCallbacks(resource_type_url);
+  }
+  Common::CallbackHandle*
+  addDynamicContextUpdateCallback(UpdateNotificationCb callback) const override {
+    return update_cb_helper_.add(callback);
+  };
 
 private:
   const xds::core::v3::ContextParams node_context_;
+  // Map from resource type URL to dynamic context parameters.
+  absl::flat_hash_map<std::string, xds::core::v3::ContextParams> dynamic_context_;
+  mutable Common::CallbackManager<absl::string_view> update_cb_helper_;
```

Comment:
I don't think we ever call remove() outside of the destructor. Consider a version of this that is fully RAII or maybe just a wrapper that calls remove?

Reply:
As discussed offline, will defer to your WiP PR on this one.